### DISCUSSION TITLE
Using HTTP2 without TLS

### DISCUSSION THREAD

**Discussion by siddhsql** [encode/httpx#3161] [General]
I am new to httpx and trying to use http2 without TLS. My server is Spring Boot running embedded Tomcat as explained [[here](https://stackoverflow.com/a/50365270/)]. I verified a sample request returns `HTTP2` using `curl`:

```
$ curl --http2-prior-knowledge -I http://localhost:8080/
HTTP/2 200
vary: Origin
vary: Access-Control-Request-Method
vary: Access-Control-Request-Headers
content-type: application/json
date: Sat, 06 Apr 2024 17:33:17 GMT
```

but when I try to use `httpx` according to [[1](https://www.python-httpx.org/http2/)] this is what I get:

```
>>> client = httpx.AsyncClient(http2=True)
>>> response = await client.get("http://localhost:8080/")
>>> response.http_version
'HTTP/1.1'
```

so it seems it not using http2. How can I get it to use http2? this is the first time I am trying httpx so apologies if I have overlooked something.

**Comment by lovelydinosaur** [comment]
To force HTTP/2 prior knowledge, use...

```python
>>> client = httpx.Client(http1=False, http2=True)
```

**Comment by siddhsql** [comment]
thanks. it seems to work:

```
>>> client = httpx.AsyncClient(http1=False, http2=True)
>>> response = await client.get("http://localhost:8080/$")
>>> response.http_version
'HTTP/2'
```

but then shouldn't this be the pattern to use? If one uses 

```
>>> client = httpx.AsyncClient(http2=True)
```

then even if server supports http2, we are using http1 which is suboptimal.

**Reply by lovelydinosaur** [reply]
HTTP/2 is typically only supported over HTTPS, not HTTP. Specifying `http2=True` ensures that the client requests either HTTP/1.1 or HTTP/2 during the SSL setup (using the ALPN protocol), and the server responds with whichever is supported.

If you want to force HTTP/2 support for a plain HTTP connection, you need to explicitly disable HTTP/1.1. ("prior knowledge" that HTTP/2 *is* supported, without requiring that info in the SSL handshake)
